[Cosmos] Cross-Partition queries, with optional external Query Engine integration

analogrelay opened the pull request:
This PR adds support for cross-partition queries. To do this, I've adjusted the syntax for the `query_items` method. As a reminder, this is the signature for that method:

```rust
pub fn query_items<T: DeserializeOwned + 'static>(
    &self,
    query: impl Into<Query>,
    partition_key: impl Into<PartitionKey>,
    options: Option<QueryOptions<'_>>,
) -> azure_core::Result<FeedPager<T>> {
```

The `partition_key` parameter accepts anything that can be converted to a `PartitionKey`. **Prior** to this PR, that meant (simplifying a bit):

* `&str`, `String`, all signed and unsigned integers, `f32`, and `f64` -> Represented a single string partition key value
* `Option<T>` where `T` is `Into<PartitionKey>` -> If the value was `Some`, then we used it as the partition key, otherwise we used an explicit `null`
* `(T1, T2)` where `T1` and `T2` are `Into<PartitionKey>` -> A two-level hierarchical PK
* `(T1, T2, T3)` where `T1`, and `T2` are `Into<PartitionKey>` -> A three-level hierarchical PK
* `()` -> An explicit `null` value. **This is where a change was made, but keep reading for more detail**

Originally, I planned to change `partition_key` so that it could accept some kind of sentinel value for a cross-partition query (`QueryPartitionStrategy::CrossPartition`), but that felt clunkier than I wanted (I suspected this when we designed the API originally but punted the problem). Then, I considered changing it to an `Option<impl Into<PartitionKey>>`, but that conflicted with support for `Option<T>` to represent potentially-null partition keys.

What I landed on is based on what we're doing in Go. A `PartitionKey` is, and has always been, a **list** of no more than 3 values. That's how it's represented in the underlying data structure, and how it's represented in the header. So, much like we did in Go, I changed our approach in Rust so that an "empty" list of partition key values indicated that a cross-partition query was desired. This means that `()`, which used to signify a single explicit `null` partition key value, now signifies an empty partition key list. A cross-partition query is thus written:

```rust
container.query_items("SELECT * FROM c", (), None)
```

This meant that we didn't have a value for explicit `null`. We still support `Option<T>::None` (where `T` is `Into<PartitionKey>`) but that makes it very clunky to specify a null literal since you have to specify a `T`, and spell out `Option::<String>::None`. The reason we support `Option<T>` is so that if you _have_ an `Option<String>` or similar in your code already, you can just pass it in. To solve that, I added `PartitionKey::NULL`, which is a `pub const` that specifies an explicit null value. So, if you want to write a query that was constrained to the partition with a `null` partition key, you write this: 

```rust
container.query_items("SELECT * FROM c", PartitionKey::NULL, None)
```

Seems reasonable, especially since I suspect `null` partition keys are quite rare.

---

Cross-partition queries are by default executed on the Gateway, which [significantly restricts query operators you can use](https://learn.microsoft.com/en-us/rest/api/cosmos-db/querying-cosmosdb-resources-using-the-rest-api#queries-that-cannot-be-served-by-gateway) to only projections and filters. However, **if** you reference `azure_data_cosmos` with the newly-added `query_engine` feature (it's behind a feature flag because it's still somewhat experimental and we may fold it in to the package entirely), you can specify `QueryOptions::query_engine` when making a query and use an externally-provided query engine (much like the support we're adding to Go). For example:

```rust
let query_engine = SomeExternalQueryEngine::new();
container.query_items("SELECT * FROM c", (), Some(QueryOptions {
    query_engine: Some(query_engine),
    ..Default::default()
}))
```

To support this, I've added a few abstractions (traits) to `azure_data_cosmos` that we'll implement in the separate Rust crate containing the query engine.

Before this feature goes in to GA, I strongly suspect we'll switch to actually _embed_ this query engine in the Rust SDK, since the query engine is itself written in Rust. Unlike other languages which will use FFI to call the query engine, there is little reason **not** to use the query engine from the Rust SDK, so folding it in as a standard dependency (`azure_data_cosmos` would depend upon the query engine crate and use it directly) seems reasonable, but for now, we're using this abstraction.

analogrelay commented on sdk/core/azure_core/src/http/pager.rs:
```diff
@@ -120,6 +120,18 @@ impl<T> PageStream<T> {
+    /// Creates a [`Pager<T>`] from a raw stream of [`Result<T>`](typespec::Result<T>) values.
+    ///
+    /// This constructor is used when you are implementing a completely custom stream and want to use it as a pager.
+    pub fn from_stream<S>(stream: S) -> Self
```
@heaths this was quite useful when I wanted to create a more complicated `PageStream` in order to support our cross-partition queries. It seems like a fairly safe addition to the public API for `PageStream`.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[azure_core](https://apiview.dev/Assemblies/Review/2af3cff65a914d06a00a34868cd180b0?revisionId=9459045c84cf482986a6d348edd52ef3)

analogrelay commented on sdk/typespec/typespec_client_core/src/http/context.rs:
```diff
@@ -15,6 +15,16 @@ pub struct Context<'a> {
 
+impl Context<'static> {
+    /// Creates a new [`Context`] that borrows from the current [`Context`].
+    pub fn into_borrowed(&self) -> Context<'_> {
```
This is a counterpart to `Context<'a>::into_owned()`. Since a `Context` holds a `Cow<'a, HashMap<...>>`, it is constrained to a specific lifetime. The `into_owned` and `into_borrowed` methods allow you to expand/contract that lifetime as necessary. The `into_owned` method clones the `HashMap` and converts it to a `Cow::Owned`, detaching it from the original `Context` that was provided. The `into_borrowed` method (which I now realize I should rename to `to_borrowed`, which I'll do in a minute) creates a new `Context<'a>` by *borrowing* the `HashMap`.

analogrelay replied:
@heaths this, and the one above in `PageStream`, are the only changes to core SDK components

heaths replied:
Given you commented on a previous commit and I was viewing all, this didn't show up for me to comment on, so see <https://github.com/Azure/azure-sdk-for-rust/pull/2577#discussion_r2080733536>. I'm fine with this in concept, but some suggestions/brainstorming on exactly impl.

heaths commented:
> So, if you want to write a query that was constrained to the partition with a null partition key

Is that how Cosmos refers to these keyless partitions then, using "null" in docs? If so, seems reasonable. Just didn't want to see "null" anywhere in Rust unless it's idiomatic to the service.

heaths commented on sdk/typespec/typespec_client_core/src/http/context.rs:
```diff
@@ -15,6 +15,16 @@ pub struct Context<'a> {
+impl Context<'static> {
+    /// Creates a new [`Context`] that borrows from the current [`Context`].
+    pub fn to_borrowed(&self) -> Context<'_> {
+        let type_map = self.type_map.as_ref();
+        Context {
+            type_map: Cow::Borrowed(type_map),
+        }
+    }
```
Given this is what `with_context` does, you might just call that or at least use the same code. Or maybe that calls this. Just want to avoid duplication as much as possible.

I'm fine adding this, though, even though they do the same thing because this is more obvious. The other was meant more for construction. Or maybe we deprecate `with_context` in favor of this one.

analogrelay replied:
That's a good point, I had actually completely missed `with_context` 🤦🏻‍♀️ . Given that we use `with_context` heavily already and it works fine for my case, I'm actually ok just removing this for now. If we do want to switch to something called as a method instead of a constructor, I'm quite happy to do that as a separate refactoring.

heaths commented on .vscode/settings.json:
```diff
@@ -7,6 +7,10 @@
   "rust-analyzer.cargo.features": "all",
   "rust-analyzer.check.command": "clippy",
+  "rust-analyzer.runnables.extraEnv": {
+    "AZURE_TEST_MODE": "live",
+    "AZURE_COSMOS_CONNECTION_STRING": "emulator"
```
Oopsies.

heaths replied:
`.vscode/launch.json` *should* be gitignore'd, so I work around these sorts of issues by just adding my own launch file, for which you can pass all the env vars you want.

analogrelay replied:
Ah yeah, good catch. Shall I gitignore it now and remove it from the repo?

analogrelay replied:
Oh, just realized this was in `settings.json`. For now I'll just remove it. The problem with `launch.json` (though it's just an aesthetics/usability problem) is that it doesn't seem to affect the "Run Test/Debug Test" inlay links on top of the test itself, which is what this setting does. You can absolutely create a `launch.json` config for running a test under the debugger with certain env vars though, it's just a bit of a personal annoyance.

Anyway, I'll just remove this for now.

heaths replied:
Right, that is a downside. But you can come close with the right selection criteria e.g.,

```sh
cargo test -p azure_data_cosmos --test test_module_name specific_test_name_pattern
```

It'd be great if VSCode had layered settings within a repo, or a way to limit user settings to a specific directory, but I'm not aware of any capability.

heaths commented on sdk/cosmos/azure_data_cosmos/Cargo.toml:
```diff
@@ -38,6 +38,7 @@ workspace = true
 [features]
 default = ["hmac_rust"]
 key_auth = [] # Enables support for key-based authentication (Primary Keys and Resource Tokens)
+query_engine = [] # Enables support for the PREVIEW external query engine
```
Should we establish a convention of prefacing "preview_" for such features? It's intentionally breaking, then, when it goes GA so people know. And before it GAs (if ever), it's much more obvious[^1] that it's preview.

/cc @RickWinter @ronniegeraghty @JeffreyRichter for thoughts.

[^1]: Especially since cargo/crates.io doesn't *yet* support feature docs.

heaths replied:
I talked with @RickWinter offline and he agrees for now: lets preface preview features with "preview_"; though, we're still open to discussion, of course.

The idea is that:

1. It's obvious it's preview whether you see the docs or not (we know how good us devs are with reading docs *beforehand*).
2. If it GAs, the customer will know for sure.
3. If it GAs, the feature might just be rolled in and crate maintainers may not want to support it as an optional feature (creates all sorts of hassles that might otherwise be unnecessary). So if you removed the feature to roll it in, they break anyway.
4. If the feature is cut and you remove it, the customer is broken anyway.

I don't really see any downside except if you keep it through GA, but then at least it's obvious that the feature is now stable.

If we're all in agreement, I'll update the guidelines.

analogrelay replied:
I'm quite happy to prefix this `preview_`, in fact I did start with that. One difference in the Rust community over others like Go, is that the idea of unstable/preview APIs in "GA" releases is accepted to some degree. Prefixing this with `preview_` would just mean there's a "breaking change" (in this **unstable** API) if the feature remains but is promoted to "GA" status. Since the feature is unstable from the start, I'd be quite happy to accept that breaking change.

I'll update it now!

heaths replied:
Whether we even allow preview features post-GA is something we need to discuss, but I'm in favor in exceptional cases, FWIW.

heaths commented on sdk/cosmos/azure_data_cosmos/src/clients/container_client.rs:
```diff
@@ -666,23 +683,38 @@ impl ContainerClient {
-        let mut base_request = Request::new(url, Method::Post);
-        let QueryPartitionStrategy::SinglePartition(partition_key) = partition_key.into();
-        base_request.insert_headers(&partition_key)?;
+        #[allow(unused_mut)] // This is used in the #[cfg(feature = "query_engine")] block
```
Maybe this instead, then? Less fragile to accidents - almost like `#[expect]` you wanted back when (and I'd still like to use).
```suggestion
        #[cfg_attr(not(feature = "query_engine"), allow(unused_mut))]
```

analogrelay replied:
Yep, I have done that in other places, I like it better. I'll do that!

heaths commented on sdk/cosmos/azure_data_cosmos/src/partition_key.rs:
```diff
@@ -95,6 +77,19 @@ impl<T: Into<PartitionKey>> From<T> for QueryPartitionStrategy {
+    /// An empty list of partition key values, which is used to signal a cross-partition query, when querying a container.
+    pub const EMPTY: PartitionKey = PartitionKey(Vec::new());
+
+    #[allow(dead_code)] // Used in the query_engine feature.
```
Same suggestion about `cfg_attr` here.

heaths commented on sdk/cosmos/azure_data_cosmos/src/partition_key.rs:
```diff
@@ -95,6 +77,19 @@ impl<T: Into<PartitionKey>> From<T> for QueryPartitionStrategy {
+
+    #[allow(dead_code)] // Used in the query_engine feature.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
```
Presumably not, but just to ask the question since I'm not an expert in Cosmos: should this also return `true` for `EMPTY`?

analogrelay replied:
This should indeed return true for `EMPTY`. `assert!(PartitionKey::EMPTY.is_empty())` would not panic.

heaths commented on sdk/cosmos/azure_data_cosmos/src/query/engine.rs:
```diff
@@ -0,0 +1,85 @@
+/// Represents a request from the query pipeline for data from a specific partition key range.
+pub struct QueryRequest {
```
Don't want a `Debug` or `SafeDebug` derive here (or elsewhere, if definitely no PII)? These are affectively models, for which we de facto derive `SafeDebug`.

analogrelay replied:
Yep, I can derive `SafeDebug`, though I may file an issue for a separate pass since I believe we don't derive `SafeDebug` in our other models today and we should.

heaths replied:
When you're sure a model wouldn't include PII, feel free to derive `Debug`.

heaths commented on sdk/cosmos/azure_data_cosmos/src/query/executor.rs:
```diff
@@ -0,0 +1,202 @@
+    // However, we don't store any T values in this, we only RETURN them.
+    // That means we use a function pointer to indicate that we don't actually operate on T directly, we just return it.
+    // Because of this, PhantomData<fn() -> T> is Send/Sync even if T isn't (see https://doc.rust-lang.org/stable/nomicon/phantom-data.html#table-of-phantomdata-patterns)
+    phantom: std::marker::PhantomData<fn() -> T>,
```
No lint on this?
```suggestion
    _phantom: std::marker::PhantomData<fn() -> T>,
```

analogrelay replied:
Nope! `PhantomData` is specifically handled in the dead-code analysis pass (https://github.com/rust-lang/rust/blob/dcecb99176edf2eec51613730937d21cdd5c8f6e/compiler/rustc_passes/src/dead.rs#L944). A field of type `PhantomData` is always considered live, even if it's never read.

heaths commented on sdk/cosmos/azure_data_cosmos/src/query/executor.rs:
```diff
@@ -0,0 +1,202 @@
+        };
+
+        // Execute the pipeline to get a page of results
+        while !pipeline.complete() {
+            // Run the pipeline
+            let results = pipeline.run()?;
```
Since `complete()` returns immediately, are we basically just taking advantage of any async waits in `pipeline.run()`? Otherwise, this seems very CPU-intensive and wasteful.

heaths reviewed (changes requested):
Mostly nits. Overall, LGTM!

analogrelay commented:
> Is that how Cosmos refers to these keyless partitions then, using "null" in docs? If so, seems reasonable. Just didn't want to see "null" anywhere in Rust unless it's idiomatic to the service.

Cosmos uses the JSON data model, so when I refer to `null` here, I'm referring to a JSON `null` literal. A `null` partition key is **not** the same as a missing partition key. It refers to the partition containing the actual JSON value `null`. For example, if your partition key is the `partitionKey` property, then this is the partition containing objects like this: `{ "partitionKey": null }`. Because this is permitted in the Cosmos data model, we need a way to represent it. In fact, I believe we also need a way to represent the JSON `undefined` value (i.e. the lack of a `"partitionKey"` property), but I'll file a bug to track that rather than jam it in here.

analogrelay commented:
I've addressed PR feedback. I don't _think_ the previous build failures were in the Cosmos crate, they appeared to be coming from `azure_keyvault_certificates` (see below), but I'll see if they reoccur after rebasing the branch.

<details><summary>Details</summary>
<p>

```
Compiling azure_security_keyvault_certificates v0.3.0 (/mnt/vss/_work/1/s/sdk/keyvault/azure_security_keyvault_certificates)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 6.11s
     Running unittests src/lib.rs (/mnt/vss/_work/1/s/target/debug/deps/azure_security_keyvault_certificates-f288530cefeeed1d)

running 6 tests
test resource::tests::canonicalizes ... ok
test resource::tests::from_certificate_bundle ... ok
test resource::tests::from_certificate_operation ... ok
test resource::tests::try_from_str ... ok
test resource::tests::test_deconstruct ... ok
test resource::tests::try_from_url ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/certificate_client.rs (/mnt/vss/_work/1/s/target/debug/deps/certificate_client-e3c673cadf979af0)

running 5 tests
test sign_jwt_with_ec_certificate ... FAILED
test purge_certificate ... FAILED
test certificate_roundtrip ... FAILED
test list_certificates ... FAILED
test update_certificate_properties ... FAILED

failures:

---- sign_jwt_with_ec_certificate stdout ----
Error: Error { context: Full(Custom { kind: HttpResponse { status: NotFound, error_code: None }, error: HttpError { status: NotFound, details: ErrorDetails { code: None, message: None }, body: {"Message":"Unable to find a record for the request POST https://sanitized.vault.azure.net/certificates/ec-certificate-signer/create?api-version=7.6-preview.2\nHeader differences:\n    \u003CAccept-Encoding\u003E is absent in request, value \u003Cgzip, deflate\u003E\nBody differences:\nRemaining Entries:\n0: https://sanitized.vault.azure.net/certificates/ec-certificate-signer/create?api-version=7.6-preview.2\n1: https://sanitized.vault.azure.net/certificates/ec-certificate-signer/pending?api-version=7.6-preview.2\n2: [https://Sanitized.vault.azure.net/keys/ec-certificate-signer//sign?api-version=7.6-preview.2\n","Status":"NotFound"},](https://sanitized.vault.azure.net/keys/ec-certificate-signer//sign?api-version=7.6-preview.2\n%22,%22Status%22:%22NotFound%22},) .. } }, "server returned error status which will not be retried: 404") }

---- purge_certificate stdout ----
Error: Error { context: Full(Custom { kind: HttpResponse { status: NotFound, error_code: None }, error: HttpError { status: NotFound, details: ErrorDetails { code: None, message: None }, body: {"Message":"Unable to find a record for the request POST https://sanitized.vault.azure.net/certificates/purge-certificate/create?api-version=7.6-preview.2\nHeader differences:\n    \u003CAccept-Encoding\u003E is absent in request, value \u003Cgzip, deflate\u003E\nBody differences:\nRemaining Entries:\n0: https://sanitized.vault.azure.net/certificates/purge-certificate/create?api-version=7.6-preview.2\n1: https://sanitized.vault.azure.net/certificates/purge-certificate/pending?api-version=7.6-preview.2\n2: https://sanitized.vault.azure.net/certificates/purge-certificate?api-version=7.6-preview.2\n3: https://sanitized.vault.azure.net/deletedcertificates/purge-certificate?api-version=7.6-preview.2\n4: https://sanitized.vault.azure.net/deletedcertificates/purge-certificate?api-version=7.6-preview.2\n5: https://sanitized.vault.azure.net/deletedcertificates/purge-certificate?api-version=7.6-preview.2\n6: [https://Sanitized.vault.azure.net/deletedcertificates/purge-certificate?api-version=7.6-preview.2\n","Status":"NotFound"},](https://sanitized.vault.azure.net/deletedcertificates/purge-certificate?api-version=7.6-preview.2\n%22,%22Status%22:%22NotFound%22},) .. } }, "server returned error status which will not be retried: 404") }

---- certificate_roundtrip stdout ----
Error: Error { context: Full(Custom { kind: HttpResponse { status: NotFound, error_code: None }, error: HttpError { status: NotFound, details: ErrorDetails { code: None, message: None }, body: {"Message":"Unable to find a record for the request POST https://sanitized.vault.azure.net/certificates/certificate-roundtrip/create?api-version=7.6-preview.2\nHeader differences:\n    \u003CAccept-Encoding\u003E is absent in request, value \u003Cgzip, deflate\u003E\nBody differences:\nRemaining Entries:\n0: https://sanitized.vault.azure.net/certificates/certificate-roundtrip/create?api-version=7.6-preview.2\n1: https://sanitized.vault.azure.net/certificates/certificate-roundtrip/pending?api-version=7.6-preview.2\n2: [https://Sanitized.vault.azure.net/certificates/certificate-roundtrip/?api-version=7.6-preview.2\n","Status":"NotFound"},](https://sanitized.vault.azure.net/certificates/certificate-roundtrip/?api-version=7.6-preview.2\n%22,%22Status%22:%22NotFound%22},) .. } }, "server returned error status which will not be retried: 404") }

---- list_certificates stdout ----
Error: Error { context: Full(Custom { kind: HttpResponse { status: NotFound, error_code: None }, error: HttpError { status: NotFound, details: ErrorDetails { code: None, message: None }, body: {"Message":"Unable to find a record for the request POST https://sanitized.vault.azure.net/certificates/list-certificates-1/create?api-version=7.6-preview.2\nHeader differences:\n    \u003CAccept-Encoding\u003E is absent in request, value \u003Cgzip, deflate\u003E\nBody differences:\nRemaining Entries:\n0: https://sanitized.vault.azure.net/certificates/list-certificates-1/create?api-version=7.6-preview.2\n1: https://sanitized.vault.azure.net/certificates/list-certificates-1/pending?api-version=7.6-preview.2\n2: https://sanitized.vault.azure.net/certificates/list-certificates-2/create?api-version=7.6-preview.2\n3: https://sanitized.vault.azure.net/certificates/list-certificates-2/pending?api-version=7.6-preview.2\n4: [https://Sanitized.vault.azure.net/certificates?api-version=7.6-preview.2\n","Status":"NotFound"},](https://sanitized.vault.azure.net/certificates?api-version=7.6-preview.2\n%22,%22Status%22:%22NotFound%22},) .. } }, "server returned error status which will not be retried: 404") }

---- update_certificate_properties stdout ----
Error: Error { context: Full(Custom { kind: HttpResponse { status: NotFound, error_code: None }, error: HttpError { status: NotFound, details: ErrorDetails { code: None, message: None }, body: {"Message":"Unable to find a record for the request POST https://sanitized.vault.azure.net/certificates/update-properties/create?api-version=7.6-preview.2\nHeader differences:\n    \u003CAccept-Encoding\u003E is absent in request, value \u003Cgzip, deflate\u003E\nBody differences:\nRemaining Entries:\n0: https://sanitized.vault.azure.net/certificates/update-properties/create?api-version=7.6-preview.2\n1: https://sanitized.vault.azure.net/certificates/update-properties/pending?api-version=7.6-preview.2\n2: https://sanitized.vault.azure.net/certificates/update-properties/?api-version=7.6-preview.2\n3: [https://Sanitized.vault.azure.net/certificates/update-properties/c1b6ef48c8f24ac5b3b9878131706dd2?api-version=7.6-preview.2\n","Status":"NotFound"},](https://sanitized.vault.azure.net/certificates/update-properties/c1b6ef48c8f24ac5b3b9878131706dd2?api-version=7.6-preview.2\n%22,%22Status%22:%22NotFound%22},) .. } }, "server returned error status which will not be retried: 404") }


failures:
    certificate_roundtrip
    list_certificates
    purge_certificate
    sign_jwt_with_ec_certificate
    update_certificate_properties

test result: FAILED. 0 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5.89s
```

</p>
</details>

heaths commented on sdk/cosmos/azure_data_cosmos/Cargo.toml:
```diff
@@ -25,7 +25,7 @@ url.workspace = true
 
 [dev-dependencies]
 azure_identity.workspace = true
-azure_core_test = { workspace = true, features = ["tracing"] }
```
You didn't want tracing for troubleshooting tests?

heaths commented on sdk/cosmos/azure_data_cosmos/src/query/executor.rs:
```diff
@@ -76,7 +76,7 @@ impl<T: DeserializeOwned + 'static> QueryExecutor<T> {
                     &self.http_pipeline,
                     &self.items_link,
-                    self.context.into_borrowed(),
+                    Context::with_context(&self.context),
```
Now that I see this, I think I am in favor of adding `into_borrowed()` (just have one call the other, perhaps) but I wouldn't hold up this PR for it. It was definitely cleaner and more intuitive the other way.

analogrelay replied:
I believe `with_context` is used in a lot of generated code, so I wanted to avoid changing it in this PR. But I agree `to_borrowed` feels cleaner

analogrelay replied:
https://github.com/Azure/azure-sdk-for-rust/issues/2591

heaths reviewed (approved):
A couple of thoughts, but nothing to hold up the PR.

heaths commented:
I see why the tests are failing. It's because of #2574 to fix #2551. "deflate" and "gzip" are no longer sent as `accept-encoding` headers. Basically, we want the default features intact but `azure_core_amqp` can't have that. I need to figure out a different way.

heaths commented:
You might want to wait for #2595.

heaths commented:
@analogrelay #2595 is in now, so you should be unblocked if you rebase.
